src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java

```diff
@@ -133,6 +129,28 @@
         this.pgStartupWait = pgStartupWait;
         Objects.requireNonNull(this.pgStartupWait, "Wait time cannot be null");
 
+        if (SystemUtils.IS_OS_LINUX) {
+            int uid = (int) new com.sun.security.auth.module.UnixSystem().getUid();
+            if (uid == 0) {
+                final List<String> command = new ArrayList<>();
+                command.addAll(Arrays.asList(
+                        "unshare", "-U",
+                        "id", "-u"
+                ));
+                final ProcessBuilder builder = new ProcessBuilder(command);
+                final Process process = builder.start();
+                BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                try {
+                    process.waitFor();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                if (process.exitValue() == 0 && br.readLine() != "0") {
```

Comment:
Unfortunately, when I tried it on Docker, it didn't work as expected. The `useUnshare` variable was always set to `false`. I also tried to run `unshare -U id -u` command manually in a docker container but I got `unshare failed: Operation not permitted` error. When I investigated the problem further, I came to the conclusion that Docker blocks the `unshare` command by default because it has caused a lot of local privilege escalations over time. So the only workaround I've found is to use `--privileged` option, which works but it doesn't seem to be an ideal solution.

Could you please investigate it further and find a suitable solution? Because without working on Docker is this change useless.

Reply:
Actually, I hadn't really targeted docker as the main use case for this change so much as throwaway testing VM's where root is the default user or for cases where one wants to test a java application that itself needs to run as root(for example some system utility than needs full privileges) on the host. This check is to ensure that `unshare` is only used when it is confirmed to function correctly so that confusing error messages don't get generated.

Basically the `unshare` command can be thought of as a lightweight docker alternative for changing the uid. I think `--privileged` is needed because this is somewhat like a [docker in docker](https://www.docker.com/blog/docker-can-now-run-within-docker/).